core/src/Streamly/Internal/Data/Stream/Reduce.hs

```diff
@@ -0,0 +1,1223 @@
+-- Copyright   : (c) 2017 Composewell Technologies
+-- License     : BSD-3-Clause
+-- Maintainer  : streamly@composewell.com
+-- Stability   : experimental
+-- Portability : GHC
+--
+-- Reduce streams by streams, folds or parsers.
+
+module Streamly.Internal.Data.Stream.Reduce
+    (
+      SessionState(..)
+    , SessionEntry(..)
+    -- * Reduce By Streams
+    , dropPrefix
+    , dropInfix
+    , dropSuffix
+
+    -- * Reduce By Folds
+    -- |
+    -- Reduce a stream by folding or parsing chunks of the stream.  Functions
+    -- generally ending in these shapes:
+    --
+    -- @
+    -- f (Fold m a b) -> Stream m a -> Stream m b
+    -- f (Parser m a b) -> Stream m a -> Stream m b
+    -- @
+
+    -- ** Generic Folding
+    -- | Apply folds on a stream.
+    , foldMany
+    , foldManyPost
+    , refoldMany
+    , foldSequence
+    , foldIterateM
+    , refoldIterateM
+
+    -- ** Chunking
+    -- | Element unaware grouping.
+    , chunksOf
+    , arraysOf
+    , intervalsOf
+
+
+    -- ** Splitting
+    -- | Streams can be sliced into segments in space or in time. We use the
+    -- term @chunk@ to refer to a spatial length of the stream (spatial window)
+    -- and the term @session@ to refer to a length in time (time window).
+
+    -- -- *** Using Element Separators
+    , splitOn
+    , splitOnSuffix
+    , splitOnPrefix
+
+    -- , splitBy
+    , splitWithSuffix
+    -- , splitByPrefix
+
+    -- -- *** Splitting By Sequences
+    , splitOnSeq
+    , splitOnSuffixSeq
+    -- , splitOnPrefixSeq
+
+    -- Keeping the delimiters
+    , splitWithSuffixSeq
+    -- , splitByPrefixSeq
+    -- , wordsBySeq
+
+    -- Splitting using multiple sequence separators
+    -- , splitOnAnySeq
+    , splitOnSuffixSeqAny
+    -- , splitOnAnyPrefixSeq
+
+    -- -- *** Splitting By Streams
+    -- -- | Splitting a stream using another stream as separator.
+
+    -- ** Keyed Window Classification
+
+    -- | Split the stream into chunks or windows by position or time. Each
+    -- window can be associated with a key, all events associated with a
+    -- particular key in the window can be folded to a single result.  The
+    -- window termination can be dynamically controlled by the fold.
+    --
+    -- The term "chunk" is used for a window defined by position of elements
+    -- and the term "session" is used for a time window.
+
+    -- *** Tumbling Windows
+    -- | A new window starts after the previous window is finished.
+
+
+
+    -- *** Keep Alive Windows
+    -- | The window size is extended if an event arrives within the specified
+    -- window size. This can represent sessions with idle or inactive timeout.
+
+
+    {-
+    -- *** Sliding Windows
+    -- | A new window starts after the specified slide from the previous
+    -- window. Therefore windows can overlap.
+    , classifySlidingChunks
+    , classifySlidingSessions
+    -- *** Sliding Window Buffers
+    -- , slidingChunkBuffer
+    -- , slidingSessionBuffer
+    -}
+
```

Comment:
We can remove line 39 to line 107 (chunking/splitting/windowing operations) for now.